### PR TITLE
Contributing.MD Changes for Adding New DCFs

### PR DESCRIPTION
**Pull request by Kpopovs1**
# [Work Item ID](./link-to-the-work-item)

## Description

This PR includes a new section in the `CONTRIBUTING.md` file detailing how to add new dev container features to the project and follow testing and documentation guidelines. This will detail a clearer and easier way for future maintainers to contribute container features to the project.

---

## PR Checklist

---

> Use the check-list below to ensure your branch is ready for PR.  If the item is not applicable, leave it blank.

- [x] I have updated the documentation accordingly.
- [x] All new and existing tests passed.
- [x] My code follows the code style of this project.
- [x] I ran the lint checks which produced no new errors nor warnings for my changes.
- [x] I have checked to ensure there aren't other open Pull Requests for the same update/change.

## Does this introduce a breaking change?

---

- [ ] Yes
- [x] No

> If this introduces a breaking change, please describe the impact and migration path for existing applications below.

## Testing

N/A

## Other information or known dependencies

N/A

### CODE REVIEW

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -176,6 +176,156 @@ The Remote - Containers extension ships into **VS Code Insiders** as new feature
...
+    1. File header and description.
+    2. Define constants and default values.
+    3. User setup and user validation.
+    4. Helper functions.
+    5. Checks for dependencies being installed or installs dependencies.
+    6. Runs container feature installs.
+    7. Gives the user correct permissions if necessary.
+    ```
+
+- Error messages from the script should echo to `STDERR`, while regular status messages should echo to `STDOUT`. This makes troubleshooting the script easier for developers. One way to facilitate this is to create an `err()` function like so:
```
Comment (juzuluag) [inline CONTRIBUTING.md:280]: Not sure if this whole section of how to "write" scripts makes sense.
This is a grey area where it depends of the shell type selected (sh, bash, ksh, zsh, ...) and the container that supports it, so some of these recommendations are specific to a shell type.

Comment (bderusha) [reply CONTRIBUTING.md:280]: Can you be more specific about which recommendations you have concerns about?

Comment (Kpopovs1) [reply CONTRIBUTING.md:280]: @juzuluag for more context, we based this on the formatting and standards of the already existing shell scripts in the repo, and general best practices followed.  If you see some points that are too specific to a shell type, would you call those out in particular? We can refactor this section accordingly.

Comment (juzuluag) [reply CONTRIBUTING.md:280]: Hi @bderusha @Kpopovs1, my comment is more towards to be flexible towards how to do things. The way I read this is like a 'forcing' howto do it. devcontainers have a lot of different containers types(i.e ubuntu, Debian, alpine), so the tooling might not always present. My suggestion is more towards examples of how to capture things (i.e stderr, catch errors) specifying that you were using xyz shell.

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -176,6 +176,156 @@ The Remote - Containers extension ships into **VS Code Insiders** as new feature
...
+        exit $EXIT_CODE
+    }
+    trap cleanup EXIT
+    ```
+
+- Consider using [shellcheck](https://github.com/koalaman/shellcheck) or the [vscode-shellcheck extension](https://github.com/vscode-shellcheck/vscode-shellcheck) to apply linting and static code analysis to the bash script to ensure it is formatted correctly.
+
+- Consider using common helper functions from [shared/utils.sh](script-library/shared/utils.sh) when managing common tasks (like updating PATH variables, or managing gpg keys) by copying them directly into your script.  
+    - NOTE: This is done to minimize the impact that any change can have on existing working scripts.
+    - Similarly, if you add a helper function to your script that could benefit others in the future, consider adding it to the `utils.sh` file as well.
```
Comment (dnastrain) [inline CONTRIBUTING.md:327]: nit: for clarity, consider repeating `shared/utils.sh` here too:

```suggestion
    - Similarly, if you add a helper function to your script that could benefit others in the future, consider adding it to the `shared/utils.sh` file as well.
```

File: CONTRIBUTING.md (hunk 1 above)
Comment (dnastrain) [reply CONTRIBUTING.md:280]: I think I see juzuluag's point -- that even though 99% of the `script-library` scripts use bash, this new doc section seems not state as much up front, and that the guidance "may vary"

**Review by dnastrain — APPROVED** [review]
overall LGTM -- knowing that one would need go through all the steps, as documented, to truly 'confirm validity'

File: CONTRIBUTING.md (hunk 1 above)
Comment (Kpopovs1) [reply CONTRIBUTING.md]: I've refactored the parts of this section that talk about specific code best practices to emphasize that they are recommendations if using bash, let me know if it's clearer now @juzuluag and @dnastrain

**Review by juzuluag — APPROVED** [review]
LGTM.

**Comment by Chuxel** [comment]
So, a note here. One of the things we're debating is whether we want to accept external contributions for features in this repository. Unlike definitions, there's already a mechanism in place to allow you to reference one from another GitHub repository. A repository labeling strategy can then be used to aid in discovery (among other UX treatments).

I almost wonder if this would be better to merge into the [staring feature template](https://github.com/microsoft/dev-container-features-template) instead and explain that this is the best way to contribute a community definition.  These directions are good for internal/official features as they are developed, but here again, there's discussion about moving these out of this repository as well since the GitHub Codespaces team will be maintaining many of them.

//cc: @2percentsilk @bamurtaugh @chrmarti @joshspicer for input since the go forward plan here is evolving. For first party features, these instructions are good.

**Comment by joshspicer** [comment]
+1 to merging something like this into the [template](https://github.com/microsoft/dev-container-features-template), which is barebones right now but will continue to develop into a natural starting point for features development

**Comment by bderusha** [comment]
Thanks for the context @Chuxel. That makes a lot of sense.  To be clear, are you suggesting that this guidance lives in this repo AND the feature template repo?  Or do you think it makes more sense to live ONLY in the feature template repo?  (or something else entirely)

**Comment by Chuxel** [comment]
There's quite a bit of conversation going on related to the future of this repository as it is getting large and we want to allow the community to contribute, without us getting in the way. At the moment there's a barrier to enter - and ideally content in other repositories could be used. We've got that now with remote features support - we still need to iron out the story for definitions. 

So, I'd rather not have the open source CONTRIBUTING.md file mention features at this time for that reason. That said,  https://github.com/microsoft/vscode-dev-containers/tree/main/script-library/container-features/README.md is the README for the internal feature set, so I think it would be fine to include there.

**Comment by Chuxel** [comment]
LGTM!  Merging!

**Pull request merged**